envoy/matcher/matcher.h

```diff
@@ -187,63 +196,67 @@ class InputMatcherFactory : public Config::TypedFactory {
 // Conceptually the data availability should start at being NotAvailable, transition to
 // MoreDataMightBeAvailable (optional, this doesn't make sense for all data) and finally
 // AllDataAvailable as the data becomes available.
-struct DataInputGetResult {
-  enum class DataAvailability {
-    // The data is not yet available.
-    NotAvailable,
-    // Some data is available, but more might arrive.
-    MoreDataMightBeAvailable,
-    // All the data is available.
-    AllDataAvailable
-  };
-
+template <class ResultDataType = std::string> struct DataInputGetResult {
   DataAvailability data_availability_;
   // The resulting data. This will be absl::nullopt if we don't have sufficient data available (as
   // per data_availability_) or because no value was extracted. For example, consider a DataInput
   // which attempts to look a key up in the map: if we don't have access to the map yet, we return
   // absl::nullopt with NotAvailable. If we have the entire map, but the key doesn't exist in the
   // map, we return absl::nullopt with AllDataAvailable.
-  absl::optional<std::string> data_;
+  absl::optional<ResultDataType> data_;
```

Comment:
How are we going to ensure that the type produced by the `input` extension is suitable for the `custom_match` extension. Also the `exact_match_map` and `prefix_match_map` expect `input` to produce a string. What happens if the input type is something other than string?